src/interfaces/ffi/hybrid_cc_aes.rs

```diff
@@ -974,90 +645,20 @@ pub unsafe extern "C" fn h_aes_decrypt(
-            *plaintext_len
-        );
-    }
-    std::slice::from_raw_parts_mut(plaintext_ptr.cast(), plaintext.len())
-        .copy_from_slice(&plaintext);
-
-    if *header_metadata_len > 0 && !decrypted_header.header_metadata.is_empty() {
-        let allocated = *header_metadata_len;
-        *header_metadata_len = decrypted_header.header_metadata.len() as c_int;
-        if allocated < *header_metadata_len {
-            ffi_bail!(
-                "The pre-allocated additional data buffer is too small; need {} bytes",
-                *header_metadata_len
-            );
-        }
-        std::slice::from_raw_parts_mut(
-            header_metadata_ptr.cast(),
-            decrypted_header.header_metadata.len(),
-        )
-        .copy_from_slice(&decrypted_header.header_metadata);
-    } else {
-        *header_metadata_len = 0;
-    }
+    ffi_write_bytes!("plaintext", &plaintext, plaintext_ptr, plaintext_len);
 
-    0
-}
-
-#[no_mangle]
-/// Convert a boolean access policy expression into a
-/// json expression that can be used to create a key using
-/// the KMIP interface
-///
-/// Returns
-///  - 0 if success
-///  - 1 in case of unrecoverable error
-///  - n if the return buffer is too small and should be of size n (including
-///    the NULL byte)
-///
-/// `json_expr_len` contains the length of the JSON string on return
-///  (including the terminating NULL byte)
-///
-/// # Safety
-pub unsafe extern "C" fn h_access_policy_expression_to_json(
-    json_expr_ptr: *mut c_char,
-    json_expr_len: *mut c_int,
-    boolean_expression_ptr: *const c_char,
-) -> c_int {
-    ffi_not_null!(json_expr_ptr, "The json_expr pointer should not be null");
-    if *json_expr_len == 0 {
-        ffi_bail!("The json_expr buffer should have a size greater than zero");
-    }
-
-    // Access policy
-    let access_policy = match CStr::from_ptr(boolean_expression_ptr).to_str() {
-        Ok(msg) => msg.to_owned(),
-        Err(_e) => {
-            set_last_error(FfiError::Generic(
-                "Hybrid Cipher: invalid attributes".to_owned(),
-            ));
-            return 1;
-        }
-    };
-    let access_policy = ffi_unwrap!(AccessPolicy::from_boolean_expression(&access_policy));
-    let json = ffi_unwrap!(serde_json::to_string(&access_policy));
-    // Build a CString
-    let cs = ffi_unwrap!(CString::new(json), "failed to convert JSON to CString");
-    // the CString as bytes
-    let bytes = cs.as_bytes_with_nul();
-    if bytes.len() > *json_expr_len as usize {
-        set_last_error(FfiError::Generic(format!(
-            "access policy to JSON: the pre-allocated buffer is too small. It should be {} bytes \
-             long",
-            bytes.len()
-        )));
-        return bytes.len() as i32;
+    if metadata_ptr.is_null() || *metadata_len == 0 {
+        let null_ptr = null_mut::<c_char>();
+        let null_len = &mut 0;
+        ffi_write_bytes!("metadata", &decrypted_header.metadata, null_ptr, null_len);
```

Comment:
Not sure to understand this part? Maybe add a comment?

Reply:
I can add a comment.